spec/re2/set_spec.rb

```diff
@@ -0,0 +1,96 @@
+RSpec.describe RE2::Set do
+  # libre2, as of 2022-06-01, in set.cc, does print to stderr (fd 2) even when
+  # :log_errors is true. So we need to temporarily redirect stderr while running
+  # tests that might print to it. The method used here is taken from the source
+  # code of MRI and seems to do the job, but it isn't guaranteed to cover all
+  # cases.
+
+  NullDevice = defined?(IO::NULL) ? IO::NULL :
+    %w[/dev/null NUL NIL: NL:].find { |dev| File.exist?(dev) }
+
+  def silence_streams(*streams)
+    on_hold = streams.collect { |stream| stream.dup }
+    streams.each do |stream|
+      stream.reopen(NullDevice)
+      stream.sync = true
+    end
+    yield
+  ensure
+    streams.each_with_index do |stream, i|
+      stream.reopen(on_hold[i])
+    end
+  end
+
+  describe "#initialize" do
+    it "returns an instance given no args" do
+      set = RE2::Set.new
+      expect(set).to be_a(RE2::Set)
+    end
+
+    it "returns an instance given only an anchor" do
+      set = RE2::Set.new(:unanchored)
+      expect(set).to be_a(RE2::Set)
+    end
+
+    it "returns an instance given an anchor and options" do
+      set = RE2::Set.new(:unanchored, :case_sensitive => false)
+      expect(set).to be_a(RE2::Set)
+    end
+
+    it "raises an error if given an inappropriate type" do
+      expect { RE2::Set.new(0) }.to raise_error(TypeError)
+    end
+  end
+
+  describe "#add" do
+    it "allows multiple patterns to be added" do
+      set = RE2::Set.new
+      expect(set.add("abc")).to eq(0)
+      expect(set.add("def")).to eq(1)
+      expect(set.add("ghi")).to eq(2)
+    end
```

Comment:
If we have multiple expectations in a single example, it might be worth using RSpec's aggregate failures to combine them into one, e.g.

```suggestion
    it "allows multiple patterns to be added", :aggregate_failures do
      set = RE2::Set.new
      expect(set.add("abc")).to eq(0)
      expect(set.add("def")).to eq(1)
      expect(set.add("ghi")).to eq(2)
    end
```